watchtower/wtclient/session_queue.go

```diff
@@ -0,0 +1,512 @@
+type sessionQueue struct {
+	started sync.Once
+	stopped sync.Once
+	forced  sync.Once
+
+	cfg *sessionQueueConfig
+
+	queue     *list.List
+	queueMtx  sync.Mutex
+	queueCond *sync.Cond
+
+	localInit *wtwire.Init
+	towerAddr *lnwire.NetAddress
+
+	quit      chan struct{}
+	forceQuit chan struct{}
+	shutdown  chan struct{}
+}
+
+// newSessionQueue intiializes a fresh sessionQueue.
+func newSessionQueue(cfg *sessionQueueConfig) *sessionQueue {
+	localInit := wtwire.NewInitMessage(
+		lnwire.NewRawFeatureVector(wtwire.WtSessionsRequired),
+		cfg.ChainHash,
+	)
+
+	towerAddr := &lnwire.NetAddress{
+		IdentityKey: cfg.ClientSession.Tower.IdentityKey,
+		Address:     cfg.ClientSession.Tower.Addresses[0],
+	}
+
+	sq := &sessionQueue{
+		cfg:       cfg,
+		queue:     list.New(),
+		localInit: localInit,
+		towerAddr: towerAddr,
+		quit:      make(chan struct{}),
+		forceQuit: make(chan struct{}),
+		shutdown:  make(chan struct{}),
+	}
+	sq.queueCond = sync.NewCond(&sq.queueMtx)
+
+	return sq
+}
+
+// Start idempotently starts the sessionQueue so that it can begin accepting
+// backups.
+func (q *sessionQueue) Start() {
+	q.started.Do(func() {
+		// TODO(conner): load prior committed state updates from disk an
+		// populate in queue.
+
+		go q.sessionManager()
+	})
+}
+
+// Stop idempotently stops the sessionQueue by initiating a clean shutdown that
+// will clear all pending tasks in the queue before returning to the caller.
+func (q *sessionQueue) Stop() {
+	q.stopped.Do(func() {
+		log.Debugf("Stopping session queue %s", q.ID())
+
+		close(q.quit)
+		q.signalUntilShutdown()
+
+		// Skip log if we also force quit.
+		select {
+		case <-q.forceQuit:
+			return
+		default:
+		}
+
+		log.Debugf("Session queue %s successfully stopped", q.ID())
+	})
+}
+
+// ForceQuit idempotently aborts any clean shutdown in progress and returns to
+// he caller after all lingering goroutines have spun down.
+func (q *sessionQueue) ForceQuit() {
+	q.forced.Do(func() {
+		log.Infof("Force quitting session queue %s", q.ID())
+
+		close(q.forceQuit)
+		q.signalUntilShutdown()
+
+		log.Infof("Session queue %s unclean shutdown complete", q.ID())
+	})
+}
+
+// ID returns the wtdb.SessionID for the queue, which can be used to uniquely
+// identify this a particular queue.
+func (q *sessionQueue) ID() *wtdb.SessionID {
+	return &q.cfg.ClientSession.ID
+}
+
+// AcceptTask attempts to queue a backupTask for delivery to the sessionQueue's
+// tower. The session will only be accepted if the queue is not already
+// exhausted and the task is successfully bound to the ClientSession.
+func (q *sessionQueue) AcceptTask(task *backupTask) (reserveStatus, bool) {
+	q.queueCond.L.Lock()
+
+	// Examine the current reserve status of the session queue.
+	curStatus := q.reserveStatus()
+	switch curStatus {
+
+	// The session queue is exhausted, and cannot accept the task because it
+	// is full. Reject the task such that it can be tried against a
+	// different session.
+	case reserveExhausted:
```

Comment:
why is that? this case should only hit if the tower has already been fully utilized

Reply:
Need to update my mental model, but at the time it seemed that if the amount of slots we needed exceeds the current available number of slots in the sessions, then rather than bin-packing up the max slots and making a new session for the remainder, we'd just make an entirely new session. If this is the case, then for the historical upload case, we might loop here forever if there're more than 65k states to be backed up?